cmd/server.go

```diff
@@ -76,51 +85,79 @@ func NewServerCommand(stopCh <-chan struct{}) *cobra.Command {
 			defer handler.Stop()
 			if snapstoreConfig == nil {
 				logger.Warnf("No snapstore storage provider configured. Will not start backup schedule.")
+				handler.Status = http.StatusOK
 				<-stopCh
 				return
 			}
-			ss, err := snapstore.GetSnapstore(snapstoreConfig)
-			if err != nil {
-				logger.Fatalf("Failed to create snapstore from configured storage provider: %v", err)
-			}
-			logger.Info("Created snapstore from provider: %s", storageProvider)
-			ssr, err := snapshotter.NewSnapshotter(
-				etcdEndpoints,
-				schedule,
-				ss,
-				logger,
-				maxBackups,
-				time.Duration(etcdConnectionTimeout),
-				certFile,
-				keyFile,
-				caFile,
-				insecureTransport,
-				insecureSkipVerify)
-			if err != nil {
-				logger.Fatalf("Failed to create snapshotter: %v", err)
-			}
 			for {
-				for {
-					logger.Infof("Probing etcd...")
-					select {
-					case <-stopCh:
-						logger.Info("Shutting down...")
-						return
-					default:
-						err = ssr.ProbeEtcd()
-					}
-					if err == nil {
-						break
+				ss, err := snapstore.GetSnapstore(snapstoreConfig)
+				if err != nil {
+					logger.Errorf("Failed to create snapstore from configured storage provider: %v", err)
+					continue
+				}
+				logger.Infof("Created snapstore from provider: %s", storageProvider)
+
+				tlsConfig := snapshotter.NewTLSConfig(
+					certFile,
+					keyFile,
+					caFile,
+					insecureTransport,
+					insecureSkipVerify,
+					etcdEndpoints)
+				ssr, err := snapshotter.NewSnapshotter(
+					schedule,
+					ss,
+					logger,
+					maxBackups,
+					time.Duration(etcdConnectionTimeout),
+					tlsConfig)
+				if err != nil {
+					logger.Errorf("Failed to create snapshotter from configured storage provider: %v", err)
+					handler.Status = http.StatusServiceUnavailable
+					continue
+				}
+
+				logger.Infof("Probing etcd...")
+				select {
+				case <-stopCh:
+					logger.Info("Shutting down...")
+					return
+				default:
+					err = ProbeEtcd(tlsConfig)
+				}
+				if err != nil {
+					logger.Errorf("Failed to probe etcd: %v", err)
+					handler.Status = http.StatusServiceUnavailable
+					continue
+				}
+
+				err = ssr.TakeFullSnapshot()
+				if err != nil {
+					if etcdErr, ok := err.(*errors.EtcdError); ok == true {
+						logger.Errorf("Snapshotter failed with etcd error: %v", etcdErr)
+
+					} else {
+						logger.Errorf("Snapshotter failed with error: %v", err)
 					}
+					handler.Status = http.StatusServiceUnavailable
+					continue
+				} else {
+					handler.Status = http.StatusOK
 				}
+
 				err = ssr.Run(stopCh)
 				if err != nil {
+					handler.Status = http.StatusServiceUnavailable
 					if etcdErr, ok := err.(*errors.EtcdError); ok == true {
-						logger.Errorf("Snapshotter failed with error: %v", etcdErr)
+						logger.Errorf("Snapshotter failed with etcd error: %v", etcdErr)
+
 					} else {
-						logger.Fatalf("Snapshotter failed with error: %v", err)
+						logger.Errorf("Snapshotter failed with error: %v", err)
```

Comment:
This should be `Fataf` right? Cause we retried with exponential backoff and still didn't succeeded. In that case it's probably because some config change, may be snapstore config like credentials or tls config and it should be reloaded. So container crash should happen.